analysis/src/machine_check.rs

```diff
@@ -280,84 +334,69 @@ impl TypeChecker {
-                                Err(e) => {
-                                    errors.extend(e);
-                                }
-                                Ok(machine) => {
-                                    res.insert(ctx.with_part(&name), Item::Machine(machine));
-                                }
-                            };
-                        }
-                        asm::SymbolValue::Import(_) => {
-                            unreachable!("Imports should have been removed")
-                        }
-                        asm::SymbolValue::Module(m) => {
-                            // add the name of this module to the context
-                            let ctx = ctx.with_part(&name);
-
-                            let m = match m {
-                                asm::Module::External(_) => unreachable!(),
-                                asm::Module::Local(m) => m,
-                            };
-
-                            match self.check_module(m, &ctx) {
-                                Err(err) => {
-                                    errors.extend(err);
-                                }
-                                Ok(m) => {
-                                    res.extend(m);
-                                }
-                            };
-                        }
-                        asm::SymbolValue::Expression(e) => {
-                            res.insert(ctx.clone().with_part(&name), Item::Expression(e));
-                        }
-                        asm::SymbolValue::TypeDeclaration(enum_decl) => {
-                            res.insert(
-                                ctx.clone().with_part(&name),
-                                Item::TypeDeclaration(enum_decl),
-                            );
-                        }
-                        asm::SymbolValue::TraitDeclaration(trait_decl) => {
-                            res.insert(
-                                ctx.clone().with_part(&name),
-                                Item::TraitDeclaration(trait_decl),
-                            );
-                        }
+                ModuleStatement::SymbolDefinition(SymbolDefinition { name, .. }) => {
+                    if let Err(e) = self.check_path(&ctx.with_part(name)) {
+                        errors.extend(e);
                     }
                 }
                 ModuleStatement::TraitImplementation(trait_impl) => {
-                    res.insert(ctx.clone(), Item::TraitImplementation(trait_impl));
+                    // comment by Thibaut: here I'm confused why we add an implementation to the items, as it is not bound to a name (here it's bound to the name of the module). Shouldn't we rather, for example, have a single `Item::Trait` and store both the declaration and the implementations there?
+                    self.insert_item(ctx, Item::TraitImplementation(trait_impl.clone()));
```

Comment:
I think the main difference is that implementations do not define symbols, so we keep them separate and at module level.

Reply:
What if we have two implementations in the same module, doesn't it lead to a collision of the keys?

Reply:
Yes, that's right, it will. I'm not an expert on the ASM part, but let me check how we can solve it.

Reply:
Ok cool, it can be a different PR. With the changes I introduce in this PR, when we see `impl Foo for Bar` it would be straightforward to add it to the `Foo` `Item::Trait` or the `Bar` `Item::?`, so that we can remove the `TraitImplementation` item entirely.

Reply:
I think I would prefer to keep them separate. Currently it makes no difference in which module the impl is, but we might make that significant (like in rust). If we re-format the asm, it will be weird that the impl suddenly moves to the module of the trait.

Reply:
I have discussed it with Chris and we believe that `ASMAnalysisFile` should include a way to handle anonymous items, but it will be included in another PR

Reply:
Also note that impls in powdr are really anonymous. There is no `impl A for B`, there is just `impl TraitName<type1, type2>{...}`. So the only sensible place to put the impls (apart from the trait) is the module where they are written.

So we could change ASMAnalysisFile to contain namedItems (the current items) and statements or directives, which are a map from String (the module name) to a vector of such statements.

Reply:
> Also note that impls in powdr are really anonymous. There is no `impl A for B`, there is just `impl TraitName<type1, type2>{...}`. So the only sensible place to put the impls (apart from the trait) is the module where they are written.

Oh I see! However how about my other idea to put the implementations in the `Item::Trait`? Is it important where an implementation is written?

> So we could change ASMAnalysisFile to contain namedItems (the current items) and statements or directives, which are a map from String (the module name) to a vector of such statements.

Then this wouldn't need to change

No strong opinion on this of course, just asking out of curiosity!

Reply:
Ah is it because trait implementations are just forwarded and end up in a PIL namespace?

Reply:
If we re-print the file, I would find it weird to move the implementation to the place where the trait was defined.

Also I could imagine other items at the module level that do not have a name, but only side-effects.

Reply:
Solved in #1725 by having a separate section for implementations (can eventually be extended to other non-declaration statement)